### PR TITLE
[Reorder] #30686 - Allow developers to filter order items for reorder

### LINKED ISSUE #30686: [Reorder feature] Feature very hard to customize
### Summary (*)

In magento version 2.4, the feature 'Reorder' changed a lot compared to 2.3. We have a new class app/code/Magento/Sales/Model/Reorder/Reorder.php with only one public method execute. Other methods are private. To customize this feature is very hard, we need to make a preference and copy/paste all methods.

### Examples (*)

One customization possible (that I actually need) is to be able to filter items for reorder. I don't want to reorder all items.

### Proposed solution

There are many possibilities to be more flexible. I'll push a PR for one idea.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [x] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by PierreLeMaguer**
### Description (*)

The new reorder feature is hard to customize for developers.
The idea here is to allow developers to filter order items to reorder thanks to a plugin on the new method app/code/Magento/Sales/Helper/Reorder.php::canReorderItem .

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/30686

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@PierreLeMaguer - It would be useful, as this may not be your only PR in this code area, to add a small description of the change made and how you envisage developers utilising the change - I'm assuming you'd expect them to create a plugin on your new method and manipulate the return? Good for searching at a later date

**Comment by PierreLeMaguer** [comment]
hello @BarnyShergold , thanks for the review. I add a small description to the PR area.

File: app/code/Magento/Sales/Helper/Reorder.php
Comment (BarnyShergold) [inline app/code/Magento/Sales/Helper/Reorder.php]: Missing return type on prototype
```suggestion
    public function canReorderItem($item, $product): bool
```

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
The Reorder.php is causing a fail on two B2B static tests. Not sure why but this is for existing code. Will need to be addressed.

Once functional tests have completed check for errors.

**Comment by PierreLeMaguer** [comment]
Hello @BarnyShergold ,

I pushed a new commit to fix static tests.

I checked the functional tests : no error here https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30687/33fbb0a2d1da4801a0d5971f86715e31/Functional/allure-report-ee/index.html#behaviors but one error here https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30687/c3bb69a2652167d4e479a67f5eaee4cf/Functional/allure-report-ce/index.html on the test  Magento\FunctionalTestingFramework.functional\MediaGalleryUiSuite  ... Not sure that it's linked to my PR.

**Comment by mbvb1223** [comment]
> Hello @BarnyShergold ,
> 
> I pushed a new commit to fix static tests.
> 
> I checked the functional tests : no error here https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30687/33fbb0a2d1da4801a0d5971f86715e31/Functional/allure-report-ee/index.html#behaviors but one error here https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30687/c3bb69a2652167d4e479a67f5eaee4cf/Functional/allure-report-ce/index.html on the test Magento\FunctionalTestingFramework.functional\MediaGalleryUiSuite ... Not sure that it's linked to my PR.

It seems the same my issue.
I also got an error 
`Magento\FunctionalTestingFramework.functional\MediaGalleryUiSuite`
 `https://studio.cucumber.io/projects/131313/test-plan/folders/1320712/scenarios/4951848: Used in pages link`

**Comment by PierreLeMaguer** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run Functional Tests CE

**Comment by BarnyShergold** [comment]
@magento run Functional Tests B2B

**Comment by BarnyShergold** [comment]
@magento run Integration Tests

**Comment by BarnyShergold** [comment]
@magento run Functional Tests B2B

**Comment by mbvb1223** [comment]
@magento run Functional Tests B2B

**Comment by romainruaud** [comment]
can you please guy tag this PR with "Partner : Smile" ? 

I used to be able to do so previously but it seems it's not the case anymore.

Regards

**Comment by PierreLeMaguer** [comment]
@magento run Functional Tests B2B

**Comment by PierreLeMaguer** [comment]
Hello @BarnyShergold , is it ok now ?
Also, is it possible to add the label Parter: Smile ?
Thanks

**Review by BarnyShergold — APPROVED** [review]
All tests have passed! I've added Smile

**Comment by magento-engcom-team** [comment]
Hi @BarnyShergold, thank you for the review.
[ENGCOM-8470](https://jira.corp.magento.com/browse/ENGCOM-8470) has been created to process this Pull Request
:eight_spoked_asterisk: @BarnyShergold, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by PierreLeMaguer** [comment]
@BarnyShergold any news about this PR ? Does it lack anything ?

**Comment by BarnyShergold** [comment]
@PierreLeMaguer I did some checking and currently the feedback is - "This PR is prioritised as P4, we are not able to process P4 PRs currently"

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
Hello @PierreLeMaguer,

Thanks for your contributions!

Moving this PR to on hold since it’s a feature request and we need PO approval for it. Once we get approval, we will proceed accordingly.

Thanks.

**Comment by engcom-Dash** [comment]
> Hello @PierreLeMaguer,
> 
> Thanks for your contributions!
> 
> Moving this PR to on hold since it’s a feature request and we need PO approval for it. Once we get approval, we will proceed accordingly.
> 
> Thanks.

Hello @PierreLeMaguer,

Thanks for your contributions!

We've received feedback from the PO, and the requested functionality—overriding the reorder feature in customizations—wasn't intended to work that way. As a result, we can't proceed further with this PR.

Thank you again for your understanding and for your contribution!

**Pull request closed without merging**
